[python] Fixed python.net.SslSocket when python-version >= 3.4 (#8401)

theJenix opened the pull request:
The issue where SslSocket throws an exception is due to hostName being null in wrap_socket.  This fix delays wrap_socket until after connect is called, so we can pass a host in.

Primary and supporting tests in tests/unit/issues/Issue8401.hx.

This also adds a test compilation in tests/unit/compile-python.hxml for testing when python-version >= 3.4, and adds the output of that to the runci/targets/Python.hx file.

theJenix committed:
[python] Fixed python.net.SslSocket when python-version >= 3.4 (#8401)

The issue where SslSocket throws an exception is due to hostName being
null in wrap_socket.  This fix delays wrap_socket until after connect is
called, so we can pass a host in.

theJenix committed:
added compilation flag to only run this for python

theJenix committed:
removed sys.net.Host import; it was unused, and it caused issues on
non-sys targets

theJenix committed:
fixed test to work across platforms (see comment in code)

fixed copy/paste error in code comments

Simn commented:
I'm fine with the change itself, but that test looks pretty awful. Why is there so much `(cast sock.__s)`? This suggests to me that the API is lacking.

theJenix commented:
> I'm fine with the change itself, but that test looks pretty awful. Why is there so much `(cast sock.__s)`? This suggests to me that the API is lacking.

It is; `python.lib.socket.Socket` is missing gettimeout, getsockopt, and getblocking (which is Python 3.7 and above) which are all part of the python socket.socket class.  I assumed there was a reason they weren't specified (hence the casts) but I can add them to the extern definition and fix the test if that would be better.  Is there anything else about the test that you want fixed?  Or was it primarily the untyped casts?

Simn commented:
I think the rest is fine. Having socket things in the unit tests can sometimes be annoying, but I suppose in this case there's not much that can go wrong.

theJenix committed:
added gettimeout, getblocking, and getsockopt to
python.lib.socket.Socket
(also added overload of setsockopt to take an Int value)

updated Issue8401 to remove casts

theJenix commented:
Cool.  I went ahead and added gettimeout, getblocking (for 3.7 and above) and getsockopt to `python.lib.socket.Socket` and updated the test to remove casts.  I also added an overload of setsockopt which is a little bit of scope creep but I thought was useful (the actual python function takes an Int instead of a Bool).  LMK if you'd prefer I take this out/put it into another PR and I can do so.

Definitely agree about having to be careful with sockets in tests.  That's why I call the private `wrapSocketWithSslContext` in the tests; python doesn't afaik do anything on the network when wrapping the SSL socket (it uses the host name to configure the socket) so I think it's ok to do in this case.

Simn commented:
Thanks, that looks much better now! Maybe a function instead of this `__s` accessing would have been good, but that's not very important.

theJenix commented:
Awesome :). Re: __s, personally I prefer to keep members like that hidden for non-test consumers of the class (it's one of the reasons I _love_ @:access in Haxe; imho solves an age old testing problem of "how do I test this private method without making it public") but I'm also happy to make that change if that's more in line with the language/library philosophy.  Let me know!

Simn commented:
Well, the obvious question is if somebody else might want to access it, but I suppose the whole idea is to abstract the native API away. And I agree that this is a good use-case for `@:access` without having to expose anything, so let's just keep it as-is.

Simn commented:
CI failures are related.

theJenix commented:
They are!  Sorry about that, I hadn't gotten back to look at this yet but I see what's up and I will make the fix.

Simn commented:
You can use `feq` instead of `eq` for float comparison.

theJenix committed:
replaced eq with feq in test for settimeout values

theJenix commented:
Thanks!  Changed two of the 3 lines to `feq` (left the test for null as an eq call).

theJenix commented:
I saw one failure on windows-test (python) when running `export/threads.py`; I don't think that's related to this change but let me know if I am missing something there and I can dive in/figure out what's up.

Simn commented:
That's unrelated indeed, the thread tests sometimes fail on python because apparently something is wrong.

Thanks!
